File: di-backend/src/main/java/com/di/service/DocumentService.java
package com.di.service;

import com.di.dto.DocumentDTO;
import com.di.dto.DocumentSearchCriteria;
import com.di.model.DocumentType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.concurrent.CompletableFuture;

public interface DocumentService {
    
    /**
     * Upload a new document
     * @param file The document file
     * @param title Document title
     * @param author Document author
     * @param username Username of the uploader
     * @return The uploaded document details
     */
    CompletableFuture<DocumentDTO> uploadDocument(MultipartFile file, String title, String author, String username);
    
    /**
     * Get a document by ID
     * @param id Document ID
     * @return The document details
     */
    DocumentDTO getDocumentById(Long id);
    
    /**
     * Delete a document by ID
     * @param id Document ID
     * @param username Username of the requester (for authorization)
     */
    void deleteDocument(Long id, String username);
    
    /**
     * Search documents by multiple criteria
     * @param criteria Search criteria
     * @param pageable Pagination and sorting information
     * @return Page of documents matching the criteria
     */
    Page<DocumentDTO> searchDocuments(DocumentSearchCriteria criteria, Pageable pageable);
    
    /**
     * Search documents by author
     * @param author Author name
     * @param pageable Pagination and sorting information
     * @return Page of documents by the author
     */
    Page<DocumentDTO> findByAuthor(String author, Pageable pageable);
    
    /**
     * Search documents by title
     * @param title Document title
     * @param pageable Pagination and sorting information
     * @return Page of documents with matching title
     */
    Page<DocumentDTO> findByTitle(String title, Pageable pageable);
    
    /**
     * Search documents by document type
     * @param documentType Document type
     * @param pageable Pagination and sorting information
     * @return Page of documents of the specified type
     */
    Page<DocumentDTO> findByDocumentType(DocumentType documentType, Pageable pageable);
    
    /**
     * Search documents by upload date range
     * @param startDate Start date
     * @param endDate End date
     * @param pageable Pagination and sorting information
     * @return Page of documents uploaded within the date range
     */
    Page<DocumentDTO> findByUploadDateBetween(LocalDateTime startDate, LocalDateTime endDate, Pageable pageable);
}

File: di-frontend/app/services/documentService.ts
import { DocumentDTO, DocumentType, PageResponse, DocumentSearchParams } from '../types/document';

const API_URL = 'http://localhost:8080/api/documents';

export async function uploadDocument(file: File, title: string, author: string, token: string): Promise<DocumentDTO> {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('title', title);
  formData.append('author', author);

  const response = await fetch(`${API_URL}/upload`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    body: formData,
  });

  if (!response.ok) {
    throw new Error('Failed to upload document');
  }

  return response.json();
}

export async function getDocumentById(id: number, token: string): Promise<DocumentDTO> {
  const response = await fetch(`${API_URL}/${id}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch document');
  }

  return response.json();
}

export async function deleteDocument(id: number, token: string): Promise<void> {
  const response = await fetch(`${API_URL}/${id}`, {
    method: 'DELETE',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to delete document');
  }
}

export async function searchDocuments(params: DocumentSearchParams, token: string): Promise<PageResponse<DocumentDTO>> {
  const queryParams = new URLSearchParams();
  
  if (params.title) queryParams.append('title', params.title);
  if (params.author) queryParams.append('author', params.author);
  if (params.documentType) queryParams.append('documentType', params.documentType);
  if (params.content) queryParams.append('content', params.content);
  if (params.startDate) queryParams.append('startDate', params.startDate);
  if (params.endDate) queryParams.append('endDate', params.endDate);
  queryParams.append('page', String(params.page || 0));
  queryParams.append('size', String(params.size || 10));

  const response = await fetch(`${API_URL}/search?${queryParams}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to search documents');
  }

  return response.json();
}